### DISCUSSION TITLE
Coco Pretraining

### DISCUSSION THREAD

**Discussion by philippneugebauer** [ultralytics/yolov5#1827] [Q&A] [answered]
So yolov5 is pretrained with COCO (https://github.com/ultralytics/yolov5/issues/1089), but I couldn't find out if it's the complete COCO dataset. Is it trained with the complete COCO dataset or is it still useful to train yolov5 with interesting parts of the COCO dataset?

**Comment by glenn-jocher** [comment, marked as answer]
@philippneugebauer the official models available for download under the release assets section are trained on COCO. They are not pretrained on anything prior to training. 

Code to reproduce is here:
https://github.com/ultralytics/yolov5#training

**Reply by philippneugebauer** [reply]
ups, I overlooked that description. I thought that was the fast start to train your model. Maybe it's worth to have a get started fast section in the readme?

**Comment by glenn-jocher** [comment]
@philippneugebauer we try to make everything as clear as possible but yes I'm sure there's room for improvement. In general if your use case is training a custom model on a small or medium dataset you'll want to start from pretrained weights:
```bash
python train.py --data custom_dataset.yaml --weights yolov5s.pt
```

If your custom dataset is large (i.e. 100k images) you may obtain better results simply training from scratch:
```bash
python train.py --data custom_dataset.yaml --weights '' --cfg yolov5s.yaml
```

See custom training tutorial to get started:
https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data

**Comment by philippneugebauer** [comment]
yeah, I figured that out myself but I did not realize the paragraph already answered my question since I thought that's general startup stuff.

Another question to the commands:

I was using `python train.py --data custom.yml --cfg yolov5s.yaml --weights 'yolov5s.pt' --batch-size 64` but afterwards yolov5 finds only my trained parameter. Is that normal or did I something wrong? As I understood it, it should also detect the stuff through the COCO dataset like persons, but it won't detect them. It detects them if I use `--weights yolov5s.pt`.

**Comment by glenn-jocher** [comment]
The model trains on all classes present in your --data, if COCO classes are not present in your data they will not be detectable by the trained model.

**Comment by philippneugebauer** [comment]
I had a misunderstanding of the `--weights` command. I thought, it's giving the results of the COCO run into the model to be created including all its classes. That's wrong?

Does it work that I have a single picture of a person and then it combines the results of pretrained dataset with mine? Or just naming that class in the `--data` file?

**Comment by glenn-jocher** [comment]
@philippneugebauer --weights provides optional pretraind weights that the model may begin training from. If your dataset contains a different class count the output layers will have different shapes and will not transfer. See PyTorch Hub tutorial for further explanation:
https://github.com/ultralytics/yolov5/issues/36


### Number of Classes
To load a pretrained YOLOv5s model with 10 output classes rather than the default 80:
```python
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True, classes=10)
```
In this case the model will be composed of pretrained weights **except for** the output layers, which are no longer the same shape as the pretrained output layers. The output layers will remain initialized by random weights.

**Comment by Gelassen** [comment]
@glenn-jocher , hello. I have a couple of questions, but instead of starting a new topic I decided to ask them here as this thread is good for fresh starters to answer unclear points. 

COCO dataset. In my first experience with YOLOv5 I succeeded with custom dataset [training](https://colab.research.google.com/drive/1SuUT9E2vEtZRHPzVB93Z9SjRkj93Gtm7?usp=sharing), but when I decided to do training on COCO dataset I faced with issue to define labeled images. COCO dataset has special format (captions, instances, persons keypoints in json) and search a proper way to interpret it. What I have found in project [repo](https://github.com/ultralytics/yolov5/blob/69be8e738f45e7908c1270eedd350f98c0c7bfa4/data/coco.yaml) just brought new questions -- what is inside those .txt files and how to convert existing 3 json annotations to this format. Could you please to share more details on how do you train yolov5 on this dataset? 

Weights. As I understand a trained model is described only by weights file (.pt extension) and together with predefined model (yolov5s.yaml) it is enough to run detect.py to spot classes in real dataset. It means I clone yolov5 repo, put my predefined weights and it is ready to detect objects. Is it correct? How to export this pre-trained weights (from training on custom dataset weights file hadn't been generated)

**Comment by glenn-jocher** [comment]
@Gelassen COCO dataset downloads automatically if requested (as the tutorials explain):
```
python train.py --data coco.yaml
```

See Train Custom Data tutorial to get started:

## Tutorials

* [Train Custom Data](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://github.com/ultralytics/yolov5/issues/475)
* [PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://github.com/ultralytics/yolov5/issues/251)
* [Test-Time Augmentation (TTA)](https://github.com/ultralytics/yolov5/issues/303)
* [Model Ensembling](https://github.com/ultralytics/yolov5/issues/318)
* [Model Pruning/Sparsity](https://github.com/ultralytics/yolov5/issues/304)
* [Hyperparameter Evolution](https://github.com/ultralytics/yolov5/issues/607)
* [Transfer Learning with Frozen Layers](https://github.com/ultralytics/yolov5/issues/1314)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Reply by MorphSeur** [reply]
Hi @glenn-jocher,

May I ask if you have a COCO dataset pretrained model?
I do not have (RAM, CPU/GPU, etc.) to train it on my own.

Regards,

**Reply by glenn-jocher** [reply]
@MorphSeur yes of course. All the latest COCO trained models are available under the latest release assets:
https://github.com/ultralytics/yolov5/releases/tag/v6.1

<img width="1625" alt="Screenshot 2022-07-07 at 13 15 06" src="https://user-images.githubusercontent.com/26833433/177760960-9572c72c-a214-4b38-bd6b-65ef6678bd9c.png">

**Reply by MorphSeur** [reply]
Dear @glenn-jocher,

Thanks a lot.
Really appreciate it.

Regards,
